<launch>

    <!-- Path for Rover's GNSS observation file -->
    <arg name="roverMeasureFile" default="" />
    
    <!-- Paths for GNSS navigation files -->
    <param name="roverMeasureFile" type="string" value="$(arg roverMeasureFile)" />
    <param name="navigationFile1" type="string" value="" />
    <param name="navigationFile2" type="string" value="" />
    <param name="navigationFile3" type="string" value="" />
    <param name="navigationFile4" type="string" value="" />
    <param name="navigationFile5" type="string" value="" />
    <param name="out_folder" type="string" value="$(find gals)/gps_lib.pos" />
    
    <!-- Optimization Parameters -->
    <arg name="verbose" default="false" /> <!-- Verbose output of g2o optimizer -->
    <arg name="windowSize" default="100" /> <!-- Size of the window for optimization -->
    <arg name="maxIterations" default="50" /> <!--Maximum iterations for each on-line optimization -->
    <arg name="maxIterationsEnd" default="50" />  <!--Maximum iterations for complete optimization (at the end) -->
    <arg name="optimizeBiasesAgain" default="true" />  <!--Whether to optimize clock biases in a window -->
    <arg name="optimizeBiasesAgainEnd" default="true" />  <!--Whether to optimize clock biases at the end -->
    <arg name="laserInform" default="10" />    <!-- Value of LiDAR information factor for LiDAR --> 
    <arg name="laserTimeOffset" default="0" /> <!-- Offset between LiDAR and GNSS timestamps -->
    <arg name="filterGPS" default="true" /> <!--Whether to filter GNSS data -->
    <arg name="maxGPSSpeed" default="20" /> <!-- Limits maximum distance between GPS poses (for filtration) -->
    <arg name="maxGPSVertSpeed" default="5" /> <!-- Limits maximum vertical distance between GPS poses (for filtration) -->
    <arg name="maxAltToDstPct" default="20" /> <!-- Maximum allowed percentage of altitude change to distance ratio (for filtration)-->
    <arg name="skipGPSPoses" default="0" />  <!-- Number of first GPS poses to skip (ie. waiting for laser)-->
    <arg name="decimation" default="10" /> <!-- Decimation of processed measurements -->
    <arg name="posesToProcess" default="1000" /> <!-- limit of poses for processing --> <!-- Must not exceed valid poses number in sequence or complete optimization (at the end) will not be called -->
    <arg name="dopplerInformFactor" default="10" /> <!-- Value of LiDAR information factor for Doppler measurements -->
    <arg name="velToDopplerRatiox10" default="20" /> <!-- Ratio of two velocities (based on GNSS poses and Doppler) x 10.0 (for filtration) -->
    <arg name="slamOdometryPath" default="/home/kcwian/Workspace/catkin_gnss_2/src/gals/dataset/UrbanNav/UrbanNav-HK-Mongkok/mapping_gps_time.txt" /> <!-- Path of SLAM system's trajectory (TUM format with timestamps converted to (week, seconds) -->
    

    <node name="gals_node" pkg="gals" type="gals_node" output="screen">

        <!--RTKLIB params-->
        <param name="mode" type="int" value="0" />  <!-- GNSS positioning mode, 0: single, 1:DGPS/DGNSS, 2: kinematic -->
        <param name="nf" type="int" value="3" />    <!-- number of frequency (1:L1,2:L1+L2,3:L1+L2+L5) -->
        <param name="soltype" type="int" value="0" />  <!-- 0:forward,1:backward,2:combined -->

        <!-- Optimization parameters -->
        <param name="verbose" type="bool" value="$(arg verbose)" />
        <param name="windowSize" type="int" value="$(arg windowSize)" />
        <param name="maxIterations" type="int" value="$(arg maxIterations)" />
        <param name="maxIterationsEnd" type="int" value="$(arg maxIterationsEnd)" />
        <param name="optimizeBiasesAgain" type="bool" value="$(arg optimizeBiasesAgain)" />
        <param name="optimizeBiasesAgainEnd" type="bool" value="$(arg optimizeBiasesAgainEnd)" />
        <param name="laserInform" type="int" value="$(arg laserInform)" />
        <param name="laserTimeOffset" type="int" value="$(arg laserTimeOffset)" />
        <param name="filterGPS" type="bool" value="$(arg filterGPS)" />
        <param name="maxGPSSpeed" type="int" value="$(arg maxGPSSpeed)" />
        <param name="maxGPSVertSpeed" type="int" value="$(arg maxGPSVertSpeed)" />
        <param name="maxAltToDstPct" type="int" value="$(arg maxAltToDstPct)" />
        <param name="skipGPSPoses" type="int" value="$(arg skipGPSPoses)" />
        <param name="decimation" type="int" value="$(arg decimation)" />
        <param name="posesToProcess" type="int" value="$(arg posesToProcess)" />
        <param name="dopplerInformFactor" type="int" value="$(arg dopplerInformFactor)" />
        <param name="velToDopplerRatiox10" type="int" value="$(arg velToDopplerRatiox10)" />
        <param name="slamOdometryPath" type="string" value="$(arg slamOdometryPath)" />
    </node>

</launch>